### DISCUSSION TITLE
Repeating content on page / in document

### DISCUSSION THREAD

**Discussion by haraman21** [QuestPDF/QuestPDF#127] [General]
Hi
I'm preparing to migrate from another pdf library to QuestPDF in my application and checking my use cases. My query seems similar to [How to Multiple Invoice Print?](https://github.com/QuestPDF/QuestPDF/discussions/14), but OP did not specify his requirement.
However in my case content repetition may be like generating multiple copies of some content in a document, either on same page, or multiple copies of a page with different header. (Header my not necessarily be Page Header, it may be a simple text component)

Example 1: Generate triplicate copies of a bill, **each on a separate page**, with only title being different (e.g. Original / Transporter Copy / Supplier Copy) and rest of the content being same on each page.
Example 2: Generating Bank/Other voucher etc. **on a single page** with 2 or 3 columns (may be a table) with only header/first row being different (e.g. Bank Copy / Depositor Copy etc.) and rest of the content/rows being same in all columns.

Currently in previous pdf library, separate objects are created for each header, and one object is created for rest of the contents which is added with each header to the document or table, twice/thrice as per requirement without any loop or duplication of code.
e.g. for a triplicate invoice each on a separate page
```
var document = ...; // create and initialize document
var Header1Object = ...; // create and initialize 1st header
var Header2Object = ...; // create and initialize 2nd header
var Header3Object = ...; // create and initialize 3rd header
var ContentObject = ...; // create and initialize content once

document.Add(Header1Object);
document.Add(ContentObject);
document.NewPage();

document.Add(Header2Object);
document.Add(ContentObject);
document.NewPage();

document.Add(Header2Object);
document.Add(ContentObject);
document.NewPage();

// or may be use add HeaderObjects, ContentObject and NewPage() in a loop to add as many copies as required
...
```
Or for a single page multi-col content
```
var document = ...; // create and initialize document
var table = ...; // create and initialize 2 col table
var Header1Object = ...; // create and initialize 1st header
var Header2Object = ...; // create and initialize 2nd header
var ContentObject = ...; // create and initialize content once

// create first row
table.Cell(Header1Object);
table.Cell(Header2Object);

// create 2nd row with same content in both columns
table.Cell(ContentObject);
table.Cell(ContentObject);

// finally add table to the document
document.Add(table);
...
```
QuestPDF being a fluent/delegate-arg pattern library, I'm not sure how to achieve the same without using loop / code repetition!

**Comment by MarcinZiabek** [comment]
Hello 😁 This is a really good question! There is an `.Element(Action<IContainer> handler)` method available ([click here](https://www.questpdf.com/documentation/api-reference.html#element)). Similarly to your code, you can abstract away all your implementation into separate methods and then just call them whenever needed. For elements repeating along multiple document types, you can also use [the Component approach](https://www.questpdf.com/documentation/patterns-and-practices.html#components). Both topics are explained in [the Getting Started tutorial](https://www.questpdf.com/documentation/getting-started.html#what-to-expect).

**Reply by haraman21** [reply]
Thanks for pointing to proper resource location and guides.
Having worked with a different library for years, the mind revolves around same pattern 🙄
I hope I can proceed with the migration sooner.

**Comment by DawidRobertNowak** [comment]
Hello,

We are closing this thread due to inactivity. Thank you for taking the time to reach out and contribute to the project — your feedback helps us improve QuestPDF.

Should you encounter any further problems, please don't hesitate to open a new issue.

Thank you!
